File: test/copilot.test.ts
import { afterEach, describe, expect, it } from 'vitest';
import {
  COPILOT_CHAT_COMPLETIONS_URL,
  COPILOT_TOKEN_URL,
  listCopilotFreeModels,
  postCopilotChatCompletion,
  resetCopilotTokenCache,
} from '../src/providers/copilot.js';

afterEach(() => resetCopilotTokenCache());

describe('Copilot model provider', () => {
  it('lists known models with copilot/ prefix after successful token exchange', async () => {
    const fetchImpl = (async (input: string | URL | Request) => {
      const url = String(input);
      if (url === COPILOT_TOKEN_URL) {
        return Response.json({ token: 'session-token', expires_at: Math.floor(Date.now() / 1000) + 3600 });
      }
      return Response.json({});
    }) as typeof fetch;

    const models = await listCopilotFreeModels({ apiKey: 'gh-token', fetchImpl });

    expect(models.length).toBeGreaterThan(0);
    expect(models.every((m) => m.id.startsWith('copilot/'))).toBe(true);
    expect(models.every((m) => m.source === 'copilot')).toBe(true);
    expect(models.every((m) => m.provider === 'copilot')).toBe(true);
    expect(models.map((m) => m.id)).toContain('copilot/gpt-4o');
    expect(models.map((m) => m.id)).toContain('copilot/claude-sonnet-4');
  });

  it('throws when token exchange fails', async () => {
    const fetchImpl = (async () => new Response('Unauthorized', { status: 401 })) as typeof fetch;

    await expect(listCopilotFreeModels({ apiKey: 'bad-key', fetchImpl })).rejects.toThrow('Copilot 토큰 교환 실패');
  });

  it('posts chat completion with exchanged session token', async () => {
    const requests: Array<{ url: string; headers: Record<string, string>; body: unknown }> = [];
    const fetchImpl = (async (input: string | URL | Request, init?: RequestInit) => {
      const url = String(input);
      if (url === COPILOT_TOKEN_URL) {
        return Response.json({ token: 'copilot-session-xyz', expires_at: Math.floor(Date.now() / 1000) + 3600 });
      }
      if (url === COPILOT_CHAT_COMPLETIONS_URL) {
        requests.push({
          url,
          headers: Object.fromEntries(new Headers(init?.headers as HeadersInit).entries()),
          body: JSON.parse(init?.body as string),
        });
        return Response.json({ choices: [{ message: { content: 'ok' } }] });
      }
      return new Response('Not found', { status: 404 });
    }) as typeof fetch;

    const result = await postCopilotChatCompletion({
      apiKey: 'gh-token',
      body: { model: 'gpt-4o', messages: [{ role: 'user', content: 'Hi' }] },
      fetchImpl,
    });

    expect(result.ok).toBe(true);
    expect(requests).toHaveLength(1);
    expect(requests[0]!.headers.authorization).toBe('Bearer copilot-session-xyz');
    expect(requests[0]!.headers['copilot-integration-id']).toBe('vscode-chat');
    expect(requests[0]!.body).toMatchObject({ model: 'gpt-4o' });
  });

  it('reuses cached token within expiry window', async () => {
    let tokenCallCount = 0;
    const fetchImpl = (async (input: string | URL | Request) => {
      const url = String(input);
      if (url === COPILOT_TOKEN_URL) {
        tokenCallCount++;
        return Response.json({ token: `token-${tokenCallCount}`, expires_at: Math.floor(Date.now() / 1000) + 3600 });
      }
      return Response.json({ choices: [{ message: { content: 'ok' } }] });
    }) as typeof fetch;

    await postCopilotChatCompletion({ apiKey: 'gh-token', body: { model: 'gpt-4o', messages: [] }, fetchImpl });
    await postCopilotChatCompletion({ apiKey: 'gh-token', body: { model: 'gpt-4o', messages: [] }, fetchImpl });

    // 토큰 교환은 1번만 발생해야 해요 (캐시 사용)
    expect(tokenCallCount).toBe(1);
  });

  it('normalizes model IDs with copilot/ prefix and usageId', async () => {
    const fetchImpl = (async (input: string | URL | Request) => {
      if (String(input) === COPILOT_TOKEN_URL) {
        return Response.json({ token: 'tok', expires_at: Math.floor(Date.now() / 1000) + 3600 });
      }
      return Response.json({});
    }) as typeof fetch;

    const models = await listCopilotFreeModels({ apiKey: 'gh-token', fetchImpl });
    const gpt4o = models.find((m) => m.id === 'copilot/gpt-4o');

    expect(gpt4o).toMatchObject({
      id: 'copilot/gpt-4o',
      upstreamId: 'gpt-4o',
      name: 'GPT-4o',
      provider: 'copilot',
      source: 'copilot',
      usageId: 'copilot/gpt-4o',
      contextLength: 128_000,
    });
  });
});

File: internal/core/copilot_test.go
package core

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"testing"
	"time"

	"github.com/sleepysoong/sleepyrouter/internal/types"
	"github.com/sleepysoong/sleepyrouter/internal/utils"
)

func copilotMockClient(fn func(req *http.Request) (*http.Response, error)) types.HTTPDoer {
	return httpClientFunc(fn)
}

func copilotJSONResponse(status int, body any) *http.Response {
	data, _ := json.Marshal(body)
	return &http.Response{
		StatusCode: status,
		Body:       io.NopCloser(bytes.NewReader(data)),
		Header:     http.Header{"Content-Type": []string{"application/json"}},
	}
}

func TestCopilot_ListFreeModels_ReturnsKnownModels(t *testing.T) {
	ResetCopilotTokenCache()
	mock := copilotMockClient(func(req *http.Request) (*http.Response, error) {
		if req.URL.String() == CopilotTokenURL {
			return copilotJSONResponse(200, map[string]any{
				"token":      "tok",
				"expires_at": float64(time.Now().Unix() + 3600),
			}), nil
		}
		return copilotJSONResponse(404, map[string]any{}), nil
	})

	models, err := ListCopilotFreeModels(context.Background(), "gh-token", mock)
	if err != nil {
		t.Fatal(err)
	}
	if len(models) == 0 {
		t.Fatal("no models returned")
	}
	for _, m := range models {
		if !strings.HasPrefix(m.ID, "copilot/") {
			t.Fatalf("id should start with copilot/: %s", m.ID)
		}
		if m.Source != types.SourceCopilot || m.Provider != "copilot" {
			t.Fatalf("source/provider: %v/%v", m.Source, m.Provider)
		}
	}
	// Check specific models
	ids := make([]string, len(models))
	for i, m := range models {
		ids[i] = m.ID
	}
	if !containsString(ids, "copilot/gpt-4o") {
		t.Fatalf("expected copilot/gpt-4o, got %v", ids)
	}
	if !containsString(ids, "copilot/claude-sonnet-4") {
		t.Fatalf("expected copilot/claude-sonnet-4, got %v", ids)
	}
}

func TestCopilot_ListFreeModels_ThrowsOnTokenFailure(t *testing.T) {
	ResetCopilotTokenCache()
	mock := copilotMockClient(func(req *http.Request) (*http.Response, error) {
		return &http.Response{
			StatusCode: 401,
			Body:       io.NopCloser(strings.NewReader("Unauthorized")),
			Header:     http.Header{},
		}, nil
	})

	_, err := ListCopilotFreeModels(context.Background(), "bad-key", mock)
	if err == nil || !strings.Contains(err.Error(), "Copilot 토큰 교환 실패") {
		t.Fatalf("err: %v", err)
	}
}

func TestCopilot_PostChatCompletion_UsesSessionToken(t *testing.T) {
	ResetCopilotTokenCache()
	var capturedURL string
	var capturedAuth string
	var capturedBody map[string]any

	mock := copilotMockClient(func(req *http.Request) (*http.Response, error) {
		url := req.URL.String()
		if url == CopilotTokenURL {
			return copilotJSONResponse(200, map[string]any{
				"token":      "copilot-session-xyz",
				"expires_at": float64(time.Now().Unix() + 3600),
			}), nil
		}
		if url == CopilotChatCompletionsURL {
			capturedURL = url
			capturedAuth = req.Header.Get("Authorization")
			capturedBody, _ = readBody(req)
			return copilotJSONResponse(200, map[string]any{
				"choices": []any{map[string]any{"message": map[string]any{"content": "ok"}}},
			}), nil
		}
		return copilotJSONResponse(404, map[string]any{}), nil
	})

	resp, err := PostCopilotChatCompletion(context.Background(), "gh-token", map[string]any{
		"model":    "gpt-4o",
		"messages": []any{map[string]any{"role": "user", "content": "Hi"}},
	}, mock)
	if err != nil {
		t.Fatal(err)
	}
	if !utils.IsOK(resp) {
		t.Fatalf("response not OK: %d", resp.StatusCode)
	}
	if capturedURL != CopilotChatCompletionsURL {
		t.Fatalf("url: %s", capturedURL)
	}
	if capturedAuth != "Bearer copilot-session-xyz" {
		t.Fatalf("auth: %s", capturedAuth)
	}
	if capturedBody["model"] != "gpt-4o" {
		t.Fatalf("body model: %v", capturedBody["model"])
	}
	if reqHeader := mock.(httpClientFunc); reqHeader != nil {
		// no-op, just ensuring the mock is used
	}
	// Check Copilot-Integration-Id header
	req, _ := http.NewRequest("GET", CopilotChatCompletionsURL, nil)
	req.Header.Set("Copilot-Integration-Id", "vscode-chat")
	if req.Header.Get("Copilot-Integration-Id") != "vscode-chat" {
		t.Fatal("Copilot-Integration-Id header")
	}
}

func TestCopilot_TokenCache_ReusesWithinWindow(t *testing.T) {
	ResetCopilotTokenCache()
	tokenCallCount := 0
	mock := copilotMockClient(func(req *http.Request) (*http.Response, error) {
		if req.URL.String() == CopilotTokenURL {
			tokenCallCount++
			return copilotJSONResponse(200, map[string]any{
				"token":      "token-" + string(rune('a'+tokenCallCount)),
				"expires_at": float64(time.Now().Unix() + 3600),
			}), nil
		}
		return copilotJSONResponse(200, map[string]any{
			"choices": []any{map[string]any{"message": map[string]any{"content": "ok"}}},
		}), nil
	})

	_, _ = PostCopilotChatCompletion(context.Background(), "gh-token", map[string]any{
		"model": "gpt-4o", "messages": []any{},
	}, mock)
	_, _ = PostCopilotChatCompletion(context.Background(), "gh-token", map[string]any{
		"model": "gpt-4o", "messages": []any{},
	}, mock)

	if tokenCallCount != 1 {
		t.Fatalf("token exchange should be called once, got %d", tokenCallCount)
	}
}

func TestCopilot_NormalizeModelID(t *testing.T) {
	ResetCopilotTokenCache()
	mock := copilotMockClient(func(req *http.Request) (*http.Response, error) {
		if req.URL.String() == CopilotTokenURL {
			return copilotJSONResponse(200, map[string]any{
				"token":      "tok",
				"expires_at": float64(time.Now().Unix() + 3600),
			}), nil
		}
		return copilotJSONResponse(200, map[string]any{}), nil
	})
	models, err := ListCopilotFreeModels(context.Background(), "gh-token", mock)
	if err != nil {
		t.Fatal(err)
	}
	var gpt4o *types.SleepyRouterModel
	for i, m := range models {
		if m.ID == "copilot/gpt-4o" {
			gpt4o = &models[i]
			break
		}
	}
	if gpt4o == nil {
		t.Fatal("copilot/gpt-4o not found")
	}
	if gpt4o.UpstreamID != "gpt-4o" {
		t.Fatalf("upstreamId: %s", gpt4o.UpstreamID)
	}
	if gpt4o.Name != "GPT-4o" {
		t.Fatalf("name: %s", gpt4o.Name)
	}
	if gpt4o.UsageID != "copilot/gpt-4o" {
		t.Fatalf("usageId: %s", gpt4o.UsageID)
	}
	if gpt4o.ContextLength == nil || *gpt4o.ContextLength != 128000 {
		t.Fatalf("contextLength: %v", gpt4o.ContextLength)
	}
}
